File: Algorithms/138. Copy List with Random Pointer/golang/solution.recursive.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

var visited map[*Node]*Node = map[*Node]*Node{}

func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}

	if node, ok := visited[head]; ok {
		return node
	}

	node := &Node{Val: head.Val}
	visited[head] = node

	node.Next = copyRandomList(head.Next)
	node.Random = copyRandomList(head.Random)
	return node
}

File: Algorithms/138. Copy List with Random Pointer/javascript/solution.recursive.js
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */

const visitedHash = new Map();

var copyRandomList = function (head) {
    if (!head) return null;
    if (visitedHash.has(head)) return visitedHash.get(head);
    const node = new Node(head.val);
    visitedHash.set(head, node);
    node.next = copyRandomList(head.next);
    node.random = copyRandomList(head.random);
    return node;
};
